### PR TITLE
Force OMP_NUM_THREADS=1 if unset

### LINKED ISSUE #6962: [RLlib] PPO torch over 5X slower than tensorflow on atari and uses up all RAM
### What is the  problem?
The recently added torch implementation of PPO #6826  is over 5X slower when training on atari (breakout) and also ends up slowly consuming all the system RAM (perf/ram_util_percent) before crashing.  Looking at logs (attached) it seems as if a possible source of the slowdown is sampler_perf/mean_inference_ms where torch is over 40ms while tensorflow is under 2ms.  Perhaps the torch PPO implementation is not utilizing the GPU nearly as much as the tensorflow version as on my systems torch is using 1394MB of GPU memory while tensorflow is using 4446MB.

I have attached logs of running both torch and tensorflow versions of PPO on atari breakout
 as well as the yaml files for the runs.

*Ray version and other system information (Python version, TensorFlow version, OS):*
System: Azure Standard_NC12 with DSVM image
python: 3.7.6
ray: 0.9.0.dev0 (built from source rev 81238945b993fe8c41331b3cf85a55129e7e4267)
OS: Ubuntu 18.04.03 LTS
torch: 1.3.0
tensorflow: 2.0.0

Note that I see similar slow speed and consumption of RAM with PPO torch on other system setups (Azure NC_6 and NC_24) and with other environments (procgen), networks (pytorch version of cnn impala network ported from baselines), and versions of torch (1.4.0).

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):
For pytorch: rllib train -v -f atari-ppo-pt.yaml
For tensorflow: rllib train -v -f atari-ppo-tf.yaml

If we cannot run your script, we cannot fix your issue.

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [X] I have verified the issue also occurs with the [latest wheels](https://ray.readthedocs.io/en/latest/installation.html).
[attachments.zip](https://github.com/ray-project/ray/files/4131405/attachments.zip)

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

PyTorch will get much slower if OMP_NUM_THREADS isn't set on many core machines. This has been an issue for other applications as well.

This PR:
- Always sets OMP_NUM_THREADS=1 on large machines (>8 cores). This prints a warning.
- If OMP_NUM_THREADS=1, warns that puts may be slow.
- Removes the auto setting of OMP_NUM_THREADS=1 in the autoscaler, since it is now done on ray import.

## Related issue number

Closes https://github.com/ray-project/ray/issues/6962

### CODE REVIEW

File: rllib/utils/framework.py [hunk 1]
```diff
@@ -87,6 +88,14 @@ def try_import_torch(error=False):
         logger.warning("Not importing Torch for test purposes.")
         return None, None
 
+    if "OMP_NUM_THREADS" not in os.environ and multiprocessing.cpu_count() > 8:
```
Comment (ericl) [inline rllib/utils/framework.py]: Avoid printing this on laptops by checking the cpu count.

File: python/ray/__init__.py [hunk 2]
```diff
@@ -10,6 +14,12 @@
                       "requires a specific version of pickle5 (which is "
                       "packaged along with Ray).")
 
+if "OMP_NUM_THREADS" not in os.environ and multiprocessing.cpu_count() > 8:
+    logger.warning("[ray] Forcing OMP_NUM_THREADS=1 to avoid performance "
```
Comment (edoakes) [inline python/ray/__init__.py]: Link to an issue/discussion?

File: python/ray/worker.py [hunk 3]
```diff
@@ -272,10 +278,22 @@ def put_object(self, value, object_id=None, pin_object=True):
...
+
         serialized_value = self.get_serialization_context().serialize(value)
-        return self.core_worker.put_serialized_object(
+        result = self.core_worker.put_serialized_object(
             serialized_value, object_id=object_id, pin_object=pin_object)
 
+        if should_warn_of_slow_puts:
+            delta = time.perf_counter() - start
+            if delta > 0.1:
+                logger.warning("OMP_NUM_THREADS=1 is set, this may slow down "
```
Comment (edoakes) [inline python/ray/worker.py]: Link here would be nice too. Also, can we do this inside of `put_serialized_object`? This is also called for task arguments in `_raylet.pyx`

**Review by edoakes — APPROVED** [review]
LGTM I would confirm with @pcmoritz if there are concerns with logging during import.

File: python/ray/worker.py (hunk 3 above)
Comment (ericl) [reply python/ray/worker.py]: Hmm I think the serialization itself may be slower.

**Comment by robertnishihara** [comment]
Some thoughts:
- Are there other environment variables that might make sense to treat similarly? E.g., `MKL_NUM_THREADS` and `OPENBLAS_NUM_THREADS`? I've needed to do the MKL one in the past to prevent a bunch of parallel tasks all using numpy from each trying to take over the machine.
- I wonder if we should just set it for workers and not the driver?
- Should we just unset it temporarily for serialization or `ray.put`? Ideally not.

**Comment by ericl** [comment]
> Are there other environment variables that might make sense to treat similarly? E.g., MKL_NUM_THREADS and OPENBLAS_NUM_THREADS? I've needed to do the MKL one in the past to prevent a bunch of parallel tasks all using numpy from each trying to take over the machine.

It seems that OMP_NUM_THREADS will also affect MKL_NUM_THREADS, so we don't need to set both.

> I wonder if we should just set it for workers and not the driver?

There are some advantages, but one issues is that any program using Tune would have their "driver" be a worker. Probably better to avoid the inconsistent behaviour, which could be very confusing to users.

> Should we just unset it temporarily for serialization or ray.put? Ideally not.

Unfortunately I don't think this actually does anything, you have to set it before importing torch / etc for it to have an effect.

**Comment by pmacalpine** [comment]
This has fixed the extreme slowdown of torch PPO I reported in #6962 on an Azure NC_12 machine with 12 cores.  I also find that setting OMP_NUM_THREADS=1 fixes the same slowdown issue on an Azure NC_6 machine with 6 cores but running 10 workers (using 'num_cpus_per_worker: 0.5').  Maybe the function as to whether OMP_NUM_THREADS should be set to 1 should not just be conditioned on the number of cores, but maybe consider the number of workers?

**Comment by snmhaines** [comment]
Has this been implemented and, if so, in which version?  I have been struggling with trying to disable hyperthreading (see recent question: https://stackoverflow.com/questions/63108084/disabling-hyperthreading-on-nodes-in-an-aws-ec2-cluster-through-ray-configuratio ).  I re-installed ray, but the problem is still there.  I tried "- export OMP_NUM_THREADS=1" in the configuration, but still no luck.

**Pull request merged**